corehq/project_limits/gauge.py

```diff
@@ -0,0 +1,188 @@
+
+
+class GaugeLimiter:
+
+    def __init__(self, gauge, get_gauge_definition):
+        """
+        :param gauge: ``Gauge`` An instance of Gauge class
+        :param get_gauge_defintions: ``function``
+            A function to get config that GaugeLimiter will use to
+            figure out throttling conditions.
+        """
+        self.gauge = gauge
+        self.gauge_definition_fn = get_gauge_definition
+
+    def wait(self):
+        if not self.allow_usage():
+            time.sleep(self.gauge_definition.throttle_for_seconds)
+
+    def allow_usage(self):
+        """
+        Checks for the throttle conditions provided in the Defintion class and evaluates them.
+        returns `True` it means that the usage should not be throttled.
+                `False` it means that the usage should be throttled.
+        """
+        if not self.gauge_definition:
+            # no throttling if config to throttle is not set.
+            return True
+        computed_value = None
+        if self.gauge_definition.actionable_metric == AVG:
+            computed_value = self.gauge.avg()
+        else:
+            computed_value = self.gauge.max()
+        return computed_value < self.gauge_definition.acceptable_value
+
+    @property
+    def gauge_definition(self):
+        """
+        Should return an instace of subclass of ThrottleDefinition
+        """
+        return self.gauge_definition_fn(self.gauge.feature_key)
+
+
+class PillowLagGaugeLimiter(GaugeLimiter):
+
+    def _has_pillow_reported_recently(self):
+        time_window = 15  # time window in minutes in which throttling should be applied
+        time_before_window = datetime.now(tz=timezone.utc) - timedelta(minutes=time_window)
+        return self.gauge.get_last_reported_time() > time_before_window
+
+    def allow_usage(self):
+        if not self.gauge_definition:
+            # No throttling if config to throttle is not set
+            return True
+        return not (self._is_ideal_throttle_condition() and self._has_pillow_reported_recently())
+
+    def _is_ideal_throttle_condition(self):
+        """
+        The ideal throttle conditions are -
+        - If throttler is enabled in the config AND
+        - If we have reported values from scopes AND
+        - The follwing conditions are considered -
+            - If both max_value and average_value are set on the Defintion model,
+            then both should be breached in order to throttle.
+            - If only max_value is defined, then only max_value would be observed.
+            - If only average_value is set, then only average value would be observed.
+        """
+        if not self.gauge_definition.is_enabled:
+            # if throttling is disabled, don't throttle
+            return False
+
+        max_observed_value = self.gauge.max()
+        avg_observed_value = self.gauge.avg()
+
+        if max_observed_value is None or avg_observed_value is None:
```

Comment:
Still trying to understand this code, so maybe not but the comment makes me think this should be:
```suggestion
        if max_observed_value is None and avg_observed_value is None:
```

Reply:
Both of these values are calculated from the same source i.e Gauge.get_values which uses cached values. Ideally if one is None then other one would be None too. But there can be a race condition which can result in one of them being None and whenever that happens we don't want to throttle as we would not have adequate values to do so.

Reply:
Ok cool. A suppose a nit would be on the comment then
> No throttling if we don't have any values

Since that gives the impression _neither_  of the values exists, which an OR will not assert. Maybe "Don't throttle unless both observed values are set"

Reply:
https://github.com/dimagi/commcare-hq/pull/34800/commits/75bb94a807a847b2290806d4cd3b24d7106e79bc